File: Days 41 - 50/Day 48/CreateTreeFromPreorderAndInorder.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class Node
	{
		public string Value { get; private set; }

		public Node Left { get; set; } = null;
		public Node Right { get; set; } = null;

		public Node(string value)
		{
			this.Value = value;
		}
	}

	internal class Day48
	{
		private static int Main(string[] args)
		{
			List<string> preorder = new List<string>() { "a", "b", "d", "e", "c", "f", "g" };
			List<string> inorder = new List<string>() { "d", "b", "e", "a", "f", "c", "g" };
			Node tree = CreateTree(preorder, inorder);

			Console.WriteLine(tree.Left.Right.Value);

			Console.ReadLine();

			return 0;
		}

		private static Node CreateTree(List<string> preorder, List<string> inorder)
		{
			if (preorder.Count == 0 || inorder.Count == 0)
			{
				return null;
			}

			Node rootNode = new Node(preorder[0]);

			if (preorder.Count == 1)
			{
				return rootNode;
			}

			for (int i = 0; i < inorder.Count; i++)
			{
				if (inorder[i] == rootNode.Value)
				{
					List<string> shorteningPreorder = preorder.GetRange(1, i);
					List<string> shorteningInorder = inorder.GetRange(0, i + 1);

					rootNode.Left = CreateTree(shorteningPreorder, shorteningInorder);

					shorteningPreorder = preorder.GetRange(i + 1, preorder.Count - i - 1);
					shorteningInorder = inorder.GetRange(i + 1, preorder.Count - i - 1);

					rootNode.Right = CreateTree(shorteningPreorder, shorteningInorder);
				}
			}

			return rootNode;
		}
	}
}

File: Days 041 - 050/Day 48/CreateTreeFromPreorderAndInorder.cpp
#include <iostream>
#include <string>
#include <vector>

struct Node
{
	std::string value;

	Node* left;
	Node* right;

	explicit Node(const std::string& value) noexcept
		: value(value), left(nullptr), right(nullptr)
	{ }

	~Node() noexcept
	{
		if (left != nullptr)
		{
			delete left;
		}

		if (right != nullptr)
		{
			delete right;
		}
	}
};

Node* CreateTree(const std::vector<std::string>& preorder, const std::vector<std::string>& inorder)
{
	if (preorder.empty() || inorder.empty())
	{
		return nullptr;
	}

	Node* rootNode = new Node(preorder[0]);

	if (preorder.size() == 1)
	{
		return rootNode;
	}

	for (unsigned int i = 0; i < inorder.size(); i++)
	{
		if (inorder[i] == rootNode->value)
		{
			std::vector<std::string> shorteningPreorder(preorder.begin() + 1, preorder.begin() + i + 1);
			std::vector<std::string> shorteningInorder(inorder.begin(), inorder.begin() + i + 1);

			rootNode->left = CreateTree(shorteningPreorder, shorteningInorder);

			shorteningPreorder = std::vector<std::string>(preorder.begin() + i + 1, preorder.end());
			shorteningInorder = std::vector<std::string>(inorder.begin() + i + 1, inorder.end());

			rootNode->right = CreateTree(shorteningPreorder, shorteningInorder);
		}
	}

	return rootNode;
}

int main(int argc, const char* argv[])
{
	Node* tree = CreateTree({ "a", "b", "d", "e", "c", "f", "g" }, { "d", "b", "e", "a", "f", "c", "g" });
	std::cout << tree->left->left->value << "\n";

	delete tree;

	std::cin.get();

	return 0;
}